stdlib/SparseArrays/src/linalg.jl

```diff
@@ -1,12 +1,32 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
-import LinearAlgebra: checksquare
+import LinearAlgebra: checksquare, sym_uplo
 using Random: rand!
 
 # In matrix-vector multiplication, the correct orientation of the vector is assumed.
 const StridedOrTriangularMatrix{T} = Union{StridedMatrix{T}, LowerTriangular{T}, UnitLowerTriangular{T}, UpperTriangular{T}, UnitUpperTriangular{T}}
 const AdjOrTransStridedOrTriangularMatrix{T} = Union{StridedOrTriangularMatrix{T},Adjoint{<:Any,<:StridedOrTriangularMatrix{T}},Transpose{<:Any,<:StridedOrTriangularMatrix{T}}}
 
+for (Wrapper, conjugation) ∈ [(:Hermitian, :adjoint), (:Symmetric, :transpose)]
+    for op ∈ [:+, :-]
+        @eval begin
+            $op(A::AbstractSparseMatrix, B::$Wrapper{<:Any,<:AbstractSparseMatrix}) = $op(A, sparse(B))
+            $op(A::$Wrapper{<:Any,<:AbstractSparseMatrix}, B::AbstractSparseMatrix) = $op(sparse(A), B)
+
+            $op(A::AbstractSparseMatrix, B::$Wrapper) = $op(A, collect(B))
+            $op(A::$Wrapper, B::AbstractSparseMatrix) = $op(collect(A), B)
+        end
+    end
+end
+for op ∈ (:+, :-)
+    @eval begin
+        $op(A::Symmetric{<:Any, <:AbstractSparseMatrix}, B::Hermitian{<:Any, <:AbstractSparseMatrix}) = $op(sparse(A), sparse(B))
+        $op(A::Hermitian{<:Any, <:AbstractSparseMatrix}, B::Symmetric{<:Any, <:AbstractSparseMatrix}) = $op(sparse(A), sparse(B))
+        $op(A::Symmetric{<:Real, <:AbstractSparseMatrix}, B::Hermitian{<:Any, <:AbstractSparseMatrix}) = $op(Hermitian(parent(A), sym_uplo(A.uplo)), B)
+        $op(A::Hermitian{<:Any, <:AbstractSparseMatrix}, B::Symmetric{<:Real, <:AbstractSparseMatrix}) = $op(A, Hermitian(parent(B), sym_uplo(B.uplo)))
+    end
+end
```

Comment:
I'm not exactly sure, but aren't these already handled by the generic `LinearAlgebra` methods above? I thought that those automatically generate sparse matrices in case both `parent`s are sparse, but the longer I look at these methods, the more they all look alike. 🤣

Reply:
Unfortunately, the methods in LinearAlgebra.jl end up causing 
```julia
::Symmetric{<:Any, <:Sparse} + ::Hermitian{<:Any, <:Sparse}
```
to be dense (not exactly sure why), so I special cased them, and then since I special cased them, that causes a method ambiguity so that then I also need to specify what to do with 
```julia
::Symmetric{<:Real, <:Sparse} + ::Hermitian{<:Real, <:Sparse}
```

Reply:
Okay, I think that's because the `Any-Any` case is not specifically handled by the above `LinearAlgebra` methods, only the `Symmetric{<:Real}` case. Then they fall back to something very generic. So I think you should be fine without the last two lines in this block?

Reply:
No, because there's an ambiguity between
```julia
 f(::Symmetric{<:Real, <:Any}) = ...
```
and
```julia
f(::Symmetric{<:Any, <:AbstractSparseMatrix}) = ... 
```
because they constraint the parameters in different ways. e.g.:
```julia
julia> using SparseArrays, LinearAlgebra; A = Symmetric(sprandn(10, 10, 0.1));

julia> f(::Symmetric{<:Real, <:Any}) = 1
f (generic function with 2 methods)

julia> f(::Symmetric{<:Any, <:AbstractSparseMatrix}) = 2
f (generic function with 2 methods)

julia> f(A)
ERROR: MethodError: f(::Symmetric{Float64,SparseMatrixCSC{Float64,Int64}}) is ambiguous. Candidates:
  f(::Symmetric{#s24,#s25} where #s25 where #s24<:Real) in Main at REPL[15]:1
  f(::Symmetric{#s26,#s27} where #s27<:(AbstractSparseArray{Tv,Ti,2} where Ti where Tv) where #s26) in Main at REPL[16]:1
Possible fix, define
  f(::Symmetric{#s26,#s27} where #s27<:(AbstractSparseArray{Tv,Ti,2} where Ti where Tv) where #s26<:Real)
Stacktrace:
 [1] top-level scope at REPL[17]:1
 [2] eval(::Module, ::Any) at ./boot.jl:331
 [3] eval_user_input(::Any, ::REPL.REPLBackend) at /home/mason/julia/usr/share/julia/stdlib/v1.4/REPL/src/REPL.jl:86
 [4] run_backend(::REPL.REPLBackend) at /home/mason/.julia/packages/Revise/C272c/src/Revise.jl:1075
 [5] top-level scope at none:0
```

Reply:
Thanks for the explanation, and sorry for the noise. 👼